src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs

```diff
@@ -35,6 +60,122 @@ internal sealed class OpenApiSchemaStore
+    public void PopulateSchemaIntoReferenceCache(OpenApiSchema schema)
+    {
+        AddOrUpdateSchemaByReference(schema);
+        if (schema.AdditionalProperties is not null)
+        {
+            AddOrUpdateSchemaByReference(schema.AdditionalProperties);
+        }
+        if (schema.Items is not null)
+        {
+            AddOrUpdateSchemaByReference(schema.Items);
+        }
+        if (schema.AllOf is not null)
+        {
+            foreach (var allOfSchema in schema.AllOf)
+            {
+                AddOrUpdateSchemaByReference(allOfSchema);
+            }
+        }
+        if (schema.AnyOf is not null)
+        {
+            foreach (var anyOfSchema in schema.AnyOf)
+            {
+                AddOrUpdateSchemaByReference(anyOfSchema);
+            }
+        }
+        if (schema.Properties is not null)
+        {
+            foreach (var property in schema.Properties.Values)
+            {
+                AddOrUpdateSchemaByReference(property);
+            }
+        }
+    }
+
+    private void AddOrUpdateSchemaByReference(OpenApiSchema schema)
+    {
+        if (_schemasWithReference.TryGetValue(schema, out var referenceId))
+        {
+            // If we've already used this reference ID else where in the document, increment a counter value to the reference
+            // ID to avoid name collisions. These collisions are most likely to occur when the same .NET type produces a different
+            // schema in the OpenAPI document because of special annotations provided on it. For example, in the two type definitions
+            // below:
+            // public class Todo
+            // {
+            //     public int Id { get; set; }
+            //     public string Name { get; set; }
+            // }
+            // public class Project
+            // {
+            //     public int Id { get; set; }
+            //     [MinLength(5)]
+            //     public string Title { get; set; }
+            // }
+            // The `Title` and `Name` properties are both strings but the `Title` property has a `minLength` annotation
+            // on it that will materialize into a different schema.
+            // {
+            //
+            //      "type": "string",
+            //      "minLength": 5
+            // }
+            // {
+            //      "type": "string"
+            // }
+            // In this case, although the reference ID  based on the .NET type we would use is `string`, the
+            // two schemas are distinct.
+            if (referenceId == null)
+            {
+                var targetReferenceId = GetSchemaReferenceId(schema);
+                if (_referenceIdCounter.TryGetValue(targetReferenceId, out var counter))
+                {
+                    counter++;
+                    _referenceIdCounter[targetReferenceId] = counter;
+                    _schemasWithReference[schema] = $"{targetReferenceId}{counter}";
```

Comment:
Is it possible for referenceId `"Foo"` and `"Foo2"` to already be incoming referenceId's? I'm thinking something along the lines of `X509Certificate` and `X509Certificate2` but DTOs. It looks like this could conflict if the type without the 2 got an incremented counter due to special annotations.

Reply:
I filed https://github.com/dotnet/aspnetcore/issues/56305 to discuss the second approach further.